### PR TITLE
fix: wait for flush to finish

### PR DESCRIPTION
**Pull request by ronag**
Refs: https://github.com/pinojs/pino/issues/1261

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -212,10 +212,9 @@ class ThreadStream extends EventEmitter {
       throw new Error('the worker is ending')
     }
 
-    if (this[kImpl].flushing && this[kImpl].buf.length + data.length >= MAX_STRING) {
```
Comment (ronag) [inline index.js:215]: Checking for flushing here won't help us to avoid MAX_STRING length on buf.

File: index.js [hunk 2]
```diff
@@ -212,10 +212,9 @@ class ThreadStream extends EventEmitter {
...
+    if (this[kImpl].buf.length + data.length >= MAX_STRING) {
       try {
         writeSync(this)
-        this[kImpl].flushing = true
```
Comment (ronag) [inline index.js:218]: This doesn't make sense?

File: index.js [hunk 3]
```diff
@@ -416,7 +415,11 @@ function writeSync (stream) {
       process.nextTick(drain, stream)
     }
   }
-  stream[kImpl].flushing = false
+
+  if (stream[kImpl].flushing) {
+    // TODO (fix): What if flushing? Wait for flushing to finish?
```
Comment (ronag) [inline index.js:420]: @mcollina Not sure how to get this...

**Comment by ronag** [comment]
@mcollina PTAL

File: index.js [hunk 4]
```diff
@@ -257,22 +261,8 @@ class ThreadStream extends EventEmitter {
...
-        return
-      }
-      if (res === 'not-equal') {
-        // TODO handle deadlock
-        this.flush(cb)
-        return
-      }
-      process.nextTick(cb)
-    })
+    writeSync(this) // TODO (fix): Make this async somehow.
```
Comment (ronag) [inline index.js:264]: https://github.com/pinojs/thread-stream/issues/56

File: index.js (hunk 2 above)
Comment (mcollina) [reply index.js:218]: It did. There is a flush scheduled in nextTick where this was correctly reset.

File: index.js [hunk 5]
```diff
@@ -485,10 +470,28 @@ function flushSync (stream) {
...
     if (++spins === 10) {
-      throw new Error('_flushSync took too long (10s)')
+      throw new Error('waitSync took too long (10s)')
     }
   }
-  // process._rawDebug('flushSync finished')
+  // process._rawDebug('waitSync finished')
+}
+
+function waitAsync (stream, cb) {
```
Comment (mcollina) [inline index.js:479]: There is an implementation of this in lib.

Comment (ronag) [reply index.js:479]: I just moved the existing code here. This is not something new.

**Comment by ronag** [comment]
@mcollina I added a test that fails on main

**Comment by mcollina** [comment]
I would need to look into this with a couple of free hours. This logic is quite tricky. I'll get to it next week.

**Comment by mcollina** [comment]
I have tried running the attached test to this PR and it passes on main.

**Comment by mcollina** [comment]
I was able to repro with:

```js
'use strict'

const t = require('tap')
const { join } = require('path')
const { file } = require('./helper')
const { createReadStream } = require('fs')
const ThreadStream = require('..')
const buffer = require('buffer')

const MAX_STRING = buffer.constants.MAX_STRING_LENGTH

t.plan(1)

const dest = file()
const stream = new ThreadStream({
  filename: join(__dirname, 'to-file.js'),
  workerData: { dest },
  sync: false
})

stream.on('close', async () => {
  t.comment('close emitted')
  let buf
  for await (const chunk of createReadStream(dest)) {
    buf = chunk
  }
  t.equal('asd', buf.toString().slice(-3))
})

stream.on('open', () => {
  t.comment('open emitted')
  stream.write('a'.repeat(MAX_STRING - 2))
  stream.write('asd')
  stream.end()
})
```

**Comment by mcollina** [comment]
My bad, I made a mistake in the previous example. We emit `'ready'` and not `'open'`.

__I cannot reproduce the problem using the attached test__.

**Comment by ronag** [comment]
You are right I had no updated my local main.

**Pull request closed without merging**